core/solver/chebyshev.cpp

```diff
@@ -0,0 +1,390 @@
+    : Chebyshev(other.get_executor())
+{
+    *this = std::move(other);
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Chebyshev<ValueType>::transpose() const
+{
+    return build()
+        .with_generated_preconditioner(
+            share(as<Transposable>(this->get_preconditioner())->transpose()))
+        .with_criteria(this->get_stop_criterion_factory())
+        .with_foci(parameters_.foci)
+        .on(this->get_executor())
+        ->generate(
+            share(as<Transposable>(this->get_system_matrix())->transpose()));
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Chebyshev<ValueType>::conj_transpose() const
+{
+    return build()
+        .with_generated_preconditioner(share(
+            as<Transposable>(this->get_preconditioner())->conj_transpose()))
+        .with_criteria(this->get_stop_criterion_factory())
+        .with_foci(conj(std::get<0>(parameters_.foci)),
+                   conj(std::get<1>(parameters_.foci)))
+        .on(this->get_executor())
+        ->generate(share(
+            as<Transposable>(this->get_system_matrix())->conj_transpose()));
+}
+
+
+template <typename ValueType>
+void Chebyshev<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    this->apply_with_initial_guess(b, x, this->get_default_initial_guess());
+}
+
+
+template <typename ValueType>
+void Chebyshev<ValueType>::apply_with_initial_guess_impl(
+    const LinOp* b, LinOp* x, initial_guess_mode guess) const
+{
+    if (!this->get_system_matrix()) {
+        return;
+    }
+    experimental::precision_dispatch_real_complex_distributed<ValueType>(
+        [this, guess](auto dense_b, auto dense_x) {
+            prepare_initial_guess(dense_b, dense_x, guess);
+            this->apply_dense_impl(dense_b, dense_x, guess);
+        },
+        b, x);
+}
+
+
+template <typename Fn>
+void visit_criteria(Fn&& fn,
+                    std::shared_ptr<const gko::stop::CriterionFactory> c)
+{
+    fn(c);
+    if (auto combined =
+            std::dynamic_pointer_cast<const stop::Combined::Factory>(c)) {
+        for (const auto& factory : combined->get_parameters().criteria) {
+            visit_criteria(std::forward<Fn>(fn), factory);
+        }
+    }
+}
+
+
+template <typename ValueType>
+template <typename VectorType>
+void Chebyshev<ValueType>::apply_dense_impl(const VectorType* dense_b,
+                                            VectorType* dense_x,
+                                            initial_guess_mode guess) const
+{
+    using Vector = matrix::Dense<ValueType>;
+    using ws = workspace_traits<Chebyshev>;
+
+    auto exec = this->get_executor();
+    this->setup_workspace();
+
+    GKO_SOLVER_VECTOR(residual, dense_b);
+    GKO_SOLVER_VECTOR(inner_solution, dense_b);
+    GKO_SOLVER_VECTOR(update_solution, dense_b);
+
+    auto old_num_max_generation = num_max_generation_;
+    // Use the scalar first
+    // get the iteration information from stopping criterion.
+    visit_criteria(
```

Comment:
It was. but @MarcelKoch suggested that getting it from stopping criterion to reduce the parameter, I think.

Reply:
If we put it as a separate parameter, then we should not allow for a stopping criteria parameter. This would be fine with me, but I think it causes other issues, as this can't be considered a solver anymore.

Reply:
The smoother doesn't have much puprpose with norm-based stopping criteria, because that would negate the benefit of avoiding reductions, right? And can it run correctly when it uses more iterations than the number of stored scalar values? If yes, then we could have two separate parameters iterations and subspace_dimension (or similar), if no, we could have just a single parameter iterations, and internally create an Iteration stopping criterion